### PR TITLE
Reset part level upon attach from disk on MergeTree

### PR DESCRIPTION
**Pull request by arthurpassos**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix too high part level reported in [#58558](https://github.com/ClickHouse/ClickHouse/issues/58558) by resetting MergeTree part levels upon attach from disk just like `ReplicatedMergeTree` [does](https://github.com/ClickHouse/ClickHouse/blob/9cd7e6155c7027baccd6dc5380d0813db94b03cc/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp#L838).

When you attach a part to a plain MergeTree table, it will be renamed to align with the part numbering in the table, adopting the same logic currently used by ReplicatedMergeTree. Previously, parts were attached 'as is', which sometimes led to issues with incorrect block numbers or part levels.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/StorageMergeTree.cpp
Comment (arthurpassos) [inline src/Storages/StorageMergeTree.cpp:2485]: Copy of `StorageMergeTree::fillNewPartName` with `level = 0`. The former is called in other places, so I am not sure it is safe to reset it there.

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit ca8726f79999d138003b354d1dcbd10fe97ad1cc with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61536/ca8726f79999d138003b354d1dcbd10fe97ad1cc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by antaljanosbenjamin — APPROVED** [review]

**Review by antaljanosbenjamin — COMMENTED** [review]
I checked a few other usages of `fillNewPartName`:
- [RMT move partition](https://pastila.nl/?000182cd/213bab83a6a21164e489c62a78ffd95a#RHx8preMNUwY2b5+N7WQIw==)
- [MT move partition](https://pastila.nl/?005f57cf/9012128a307d41b59301357b06d80833#Rb8+WcKT3Yma3p4bCwO/bQ==)
- [RMT replace partition from](https://pastila.nl/?000a9f86/52316906bf11a566febc42daf1be6357#lLn1iCrWcrRA5XZl7Wf0eg==)
- [MT replace partition from](https://pastila.nl/?006d2461/81cbd80ef4fd25ce6cc9337854e45a40#k/S2qa6Tc0fPg7khOdx4rQ==)

My conclusion is in these cases for both RMT and MT the "mutation" is reset, but not the level. I don't know if that is good or not, but at least it consistent between MT and RMT. The only other usage of `fillNewPartName` is in `attachRestoredParts`, that I haven't tested. So I think it is better leave the other function as is.

**Comment by alexey-milovidov** [comment]
@antaljanosbenjamin The category looks wrong, see
> Bugfix validation — New test(s) failed to reproduce a bug

You can change to "Improvement"

File: tests/queries/0_stateless/03013_test_part_level_is_reset_attach_from_disk_mt.sql
Comment (antaljanosbenjamin) [inline tests/queries/0_stateless/03013_test_part_level_is_reset_attach_from_disk_mt.sql:1]:
```suggestion
SET alter_sync = 2;
-- {echoOn}
```
After going over second time, I think this can be useful. In some wierd cases this test might be flaky if the server slow to optimize/detach/attach parts. By explicitly waiting on these operations to finish the test can be more robuts.

Comment (arthurpassos) [reply tests/queries/0_stateless/03013_test_part_level_is_reset_attach_from_disk_mt.sql:2]: Done

**Comment by arthurpassos** [comment]
I'll change the description to improvement. Adding a test that reproduces the issue to the PR is tricky, it requires manual intervention in data part names using the filesystem. That's why I did not add them.

I have written a script that reproduces that crash: https://pastebin.com/1vx7iyC2. With these changes, it does not happen anymore.

I haven't had the time to do any further testing, tho. Do you think we should test some other scenarios before merging this PR?

**Comment by antaljanosbenjamin** [comment]
> I'll change the description to improvement. Adding a test that reproduces the issue to the PR is tricky, it requires manual intervention in data part names using the filesystem. That's why I did not add them.

I think you misunderstood the situation. I think the bug is not resetting the level of attached parts. That means the crash is only a sympton of the bug. I haven't checked your test, but I have very high confidence based on what I understood from your repro, my investigation, etc. that if you run your test against the current master, if will fail. It won't crash, but the level of the attached parts won't reset. Therefore I think bugfix is a completely fine category.

The bugfix validation test failed because of some issues in our CI [ClickHouse couldn't lunch](https://s3.amazonaws.com/clickhouse-test-reports/61536/1216e708f1a515b5aa0d9755affda13f138ddcba/bugfix_validation/clickhouse-server__stateless_tests_release.err.log) in that specific configuration, resulting an invalid `check_status.tsv`. As a result our CI logic didn't find your new test (and you haven't added or modified any integration tests) so it reported that it couldn't reproduce the bug (which is technically true, but for a bad reason). 

All that said, I think the right thing is to keep it as a bugfix and backport it to older version. If we keep it as an improvement, then it is hard to reason why to backport it.

**Comment by arthurpassos** [comment]
> > I'll change the description to improvement. Adding a test that reproduces the issue to the PR is tricky, it requires manual intervention in data part names using the filesystem. That's why I did not add them.
> 
> I think you misunderstood the situation. I think the bug is not resetting the level of attached parts. That means the crash is only a sympton of the bug. I haven't checked your test, but I have very high confidence based on what I understood from your repro, my investigation, etc. that if you run your test against the current master, if will fail. It won't crash, but the level of the attached parts won't reset. Therefore I think bugfix is a completely fine category.
> 
> The bugfix validation test failed because of some issues in our CI [ClickHouse couldn't lunch](https://s3.amazonaws.com/clickhouse-test-reports/61536/1216e708f1a515b5aa0d9755affda13f138ddcba/bugfix_validation/clickhouse-server__stateless_tests_release.err.log) in that specific configuration, resulting an invalid `check_status.tsv`. As a result our CI logic didn't find your new test (and you haven't added or modified any integration tests) so it reported that it couldn't reproduce the bug (which is technically true, but for a bad reason).
> 
> All that said, I think the right thing is to keep it as a bugfix and backport it to older version. If we keep it as an improvement, then it is hard to reason why to backport it.
> 
> P.S.: I have run the test with a close to master version and [it failed](https://pastila.nl/?0090a1fa/acb33f026a0396e3d5e60d99794cb40c#EgKESwXr7mVBvv/yAh/CkA==) just as it is expected.

Ah, I see. Thanks for the explanation :). Moved it back to `Bug fix`.

**Comment by arthurpassos** [comment]
@antaljanosbenjamin We'll only be able to QA this in ~2 weeks. Do you want to wait until then or you think it is safe / better to merge it as is?

**Comment by antaljanosbenjamin** [comment]
It looks good, I was waiting for some additional checks to run. I will merge it.

File: tests/queries/0_stateless/03013_test_part_level_is_reset_attach_from_disk_mt.sql
Comment (antaljanosbenjamin) [inline tests/queries/0_stateless/03013_test_part_level_is_reset_attach_from_disk_mt.sql:1]:
```suggestion
-- Tags: no-shared-merge-tree
SET alter_sync = 2;
```

**Review by antaljanosbenjamin — APPROVED** [review]
Actually I was wrong. This is kind of necessary because SharedMergeTree has different block numbers.

**Comment by antaljanosbenjamin** [comment]
I merged master because of a failing test.

**Comment by arthurpassos** [comment]
> I merged master because of a failing test.

Thanks :+1:

**Comment by antaljanosbenjamin** [comment]
The tests are flaky, the fuzzer found an already known bug and performance tests failed because of some issues in the our CI I think it is very unlikely that this change affects performance at all. Merging.

**Comment by wzb5212** [comment]
@antaljanosbenjamin Like attach partition from another table, move partition, replace partition, whether it is necessary to set level to 0 ?

```
ALTER TABLE table2 [ON CLUSTER cluster] ATTACH PARTITION partition_expr FROM table1
ALTER TABLE table2 [ON CLUSTER cluster] REPLACE PARTITION partition_expr FROM table1
ALTER TABLE table_source [ON CLUSTER cluster] MOVE PARTITION partition_expr TO TABLE table_dest
```

**Pull request merged**
